text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,328 @@
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+-->
+
+For a single project, it is possible to use the `CARGO_TARGET_DIR` environment variable (or the `target-dir` TOML config option or the `--target-dir` command-line flag) to change the position of the `target/` directory used for build artifacts during compilation with Cargo.
+
+While this option is useful for single-project environments (simple CI builds, builds through other build systems like Meson or Bazel), in multi-projects environment, like personal machines or repos with multiple workspaces, it conflates every build directory under the configured path: `CARGO_TARGET_DIR` directly replaces the `<workspace>/target/` directory.
+
+`CARGO_TARGET_DIRECTORIES` (or the `target-directories` TOML option or the `--target-directories` command-line flag) instead acts as a parent for those `target` directories.
+
+Below is an example of the behavior with `CARGO_TARGET_DIR` versus the one with `CARGO_TARGET_DIRECTORIES`:
+
+## Example
+
+Consider this directory tree:
+
+```text
+/Users/
+├─ poliorcetics/
+│  ├─ work/
+│  │  ├─ work-project/
+│  │  │  ├─ Cargo.toml
+│  │  │  ├─ crate-1/
+│  │  │  │  ├─ Cargo.toml
+│  │  │  ├─ crate-2/
+│  │  │  │  ├─ Cargo.toml
+│  ├─ perso/
+│  │  ├─ perso-1/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ perso-2/
+│  │  │  ├─ Cargo.toml
+
+/cargo-cache/
+```
+
+#### With `CARGO_TARGET_DIR=/cargo-cache`
+
+`cd /Users/poliorcetics/work/work-project && cargo build` produces artifacts directly in `/cargo-cache/debug/...`
+
+A subsequent `cargo build` in `project-1` will work with the same artifact, potentially having conflicting features for dependencies for example.
+
+A `cargo clean` will delete the entire `/cargo-cache` directory, for all projects at once.
+
+It's possible to produce invalid state in the target dir by having unrelated projects writing in the same place.
+
+It's not possible to have to projects building at once because Cargo locks its target directory during builds.
+
+#### With `CARGO_TARGET_DIRECTORIES=/cargo-cache`
+
+`cd /Users/poliorcetics/work/work-project && cargo build` produces artifacts in `/cargo-cache/work-project/debug/...`
```

Comment:
These examples need to be updated for hashing

Reply:
Yes, done